lib/ash_json_api/includes/includer.ex

```diff
@@ -71,12 +71,37 @@ defmodule AshJsonApi.Includes.Includer do
     end)
   end
 
+  defp merge_linkages(record, to_merge) do
+    linkage_to_merge = Map.get(to_merge, :__linkage__, %{})
+
+    record
+    |> Map.update!(:__linkage__, fn linkage ->
+      Map.merge(linkage, linkage_to_merge, fn _key, a, b ->
+        a ++ b
+      end)
+    end)
+  end
+
   defp flatten_includes_list({related, includes_list}) do
-    {related,
-     includes_list
-     |> List.flatten()
-     |> Enum.uniq_by(
-       &{AshJsonApi.Resource.encode_primary_key(&1), AshJsonApi.Resource.Info.type(&1)}
-     )}
+    includes =
```

Comment:
I think we can make this all even more efficient (and remove the `flatten` step) if we do this reducer as we're building up the include list. This prevents us from ever having to make the large lists of results.

Reply:
🤔 hmm, I'll be honest, I'm not quite following what that would look like. Are you saying we should be building up this map as we iterate over the relationships in the `get_includes_list/2` function?

Reply:
Yep! So this:

```elixir
defp get_includes_list(preloaded, include_keyword) do
    include_keyword
    |> Enum.reduce({preloaded, []}, fn {relationship, further},
                                       {preloaded_without_linkage, includes_list} ->
      {related, further_includes} =
        preloaded
        |> Enum.flat_map(fn record ->
          record
          |> Map.get(relationship, [])
          |> List.wrap()
        end)
        |> get_includes_list(further)

      preloaded_with_linkage =
        Enum.map(
          preloaded_without_linkage,
          fn record ->
            related =
              record
              |> Map.get(relationship, [])
              |> List.wrap()
            add_linkage(record, relationship, related)
          end
        )
      {preloaded_with_linkage, [related, further_includes, includes_list]}
    end)
    |> flatten_includes_list()
  end
```
instead of flattening, would be a recursive reducer